### PR TITLE
More botany logging

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request
as requested by Ruko

Adds logging to:

- placing punji sticks
- hitting someone with a plant that teleports, injects chemicals or shocks
- slipping on plants with those traits
- getting injected by a plant now loggs the chemicals
- plants with smokes are now admin message and are added to botany.log

## Why It's Good For The Game

more logs

## Changelog
:cl:
admin: More botany logging.
/:cl:

### CODE REVIEW

File: code/modules/hydroponics/plant_genes.dm [hunk 1]
```diff
@@ -253,16 +253,31 @@
...
 /datum/plant_gene/trait/cell_charge/on_slip(obj/item/reagent_containers/food/snacks/grown/G, mob/living/carbon/C)
-	var/power = G.seed.potency*rate
+	var/power = round(G.seed.potency*rate)
 	if(prob(power))
-		C.electrocute_act(round(power), G, 1, 1)
+		var/offender = G.fingerprintslast
+		C.electrocute_act(power, G, 1, 1)
+		var/turf/T = get_turf(C)
+		if(C.ckey != offender)
+			C.investigate_log("[C] has slipped on electric plant at [AREACOORD(T)] 	 last touched by [offender].", INVESTIGATE_BOTANY)
```
Comment (ghost) [inline code/modules/hydroponics/plant_genes.dm]:
```suggestion
			C.investigate_log("[C] has slipped on an electric plant at [AREACOORD(T)] 	 last touched by [offender].", INVESTIGATE_BOTANY)
```

File: code/modules/hydroponics/plant_genes.dm [hunk 2]
```diff
@@ -330,14 +345,30 @@
...
+		if(C.ckey == offender)		//what's the point of logging someone attacking himself
+			return
+		var/mob/living/carbon/badperson
+		for(var/mob/M in GLOB.mob_living_list)
+			var/mob/living/carbon/K = M
+			if(K.ckey == offender)
+				badperson = K
+				break
+		log_combat(badperson, C, "hit", G, "at [AREACOORD(T)] teleporting him to [AREACOORD(C)]")
+		C.investigate_log("has been hit by a bluespace plant at [AREACOORD(T)] teleporting him to [AREACOORD(C)] last touched by [offender].", INVESTIGATE_BOTANY)
```
Comment (ghost) [inline code/modules/hydroponics/plant_genes.dm]:
```suggestion
		C.investigate_log("has been hit by a bluespace plant at [AREACOORD(T)] teleporting them to [AREACOORD(C)] last touched by [offender].", INVESTIGATE_BOTANY)
```

File: code/modules/hydroponics/plant_genes.dm [hunk 3]
```diff
@@ -330,14 +345,30 @@
...
 
 /datum/plant_gene/trait/teleport/on_slip(obj/item/reagent_containers/food/snacks/grown/G, mob/living/carbon/C)
 	var/teleport_radius = max(round(G.seed.potency / 10), 1)
 	var/turf/T = get_turf(C)
+	var/offender = G.fingerprintslast
 	to_chat(C, "<span class='warning'>You slip through spacetime!</span>")
 	do_teleport(C, T, teleport_radius, channel = TELEPORT_CHANNEL_BLUESPACE)
+	if(C.ckey != offender)			//what's the point of logging someone attacking himself
+		C.investigate_log("[C] has slipped on bluespace plant at [AREACOORD(T)] teleporting him to [AREACOORD(C)] last touched by [offender].", INVESTIGATE_BOTANY)
```
Comment (ghost) [inline code/modules/hydroponics/plant_genes.dm]:
```suggestion
		C.investigate_log("[C] has slipped on bluespace plant at [AREACOORD(T)] teleporting them to [AREACOORD(C)] last touched by [offender].", INVESTIGATE_BOTANY)
```

File: code/modules/hydroponics/plant_genes.dm [hunk 4]
```diff
@@ -330,14 +345,30 @@
...
+		log_combat(C, G, "slipped on", null, "teleporting him from [AREACOORD(T)] to [AREACOORD(C)] last touched by [offender].")
```
Comment (ghost) [inline code/modules/hydroponics/plant_genes.dm]:
```suggestion
		log_combat(C, G, "slipped on", null, "teleporting them from [AREACOORD(T)] to [AREACOORD(C)] last touched by [offender].")
```

File: code/modules/hydroponics/plant_genes.dm [hunk 5]
```diff
@@ -419,6 +462,17 @@
...
+			var/offender = G.fingerprintslast
+			var/mob/living/carbon/badperson
+			if(L.ckey != offender)			//what's the point of logging someone attacking himself
+				var/turf/T = get_turf(L)
+				for(var/mob/M in GLOB.mob_living_list)
+					var/mob/living/carbon/C = M
+					if(C.ckey == offender)
+						badperson = C
+						break
+				log_combat(badperson, L, "hit", G, "at [AREACOORD(T)] injecting him with [G.reagents.log_list()]")
```
Comment (ghost) [inline code/modules/hydroponics/plant_genes.dm]:
```suggestion
				log_combat(badperson, L, "hit", G, "at [AREACOORD(T)] injecting them with [G.reagents.log_list()]")
```

File: code/modules/hydroponics/plant_genes.dm [hunk 6]
```diff
@@ -419,6 +462,17 @@
...
+				L.investigate_log("[L] has been prickled by a plant at [AREACOORD(T)] injecting him with [G.reagents.log_list()] last touched by [offender].", INVESTIGATE_BOTANY)
```
Comment (ghost) [inline code/modules/hydroponics/plant_genes.dm]:
```suggestion
				L.investigate_log("[L] has been prickled by a plant at [AREACOORD(T)] injecting them with [G.reagents.log_list()] last touched by [offender].", INVESTIGATE_BOTANY)
```

**Review by ghost — DISMISSED** [review]

**Comment by qwertyquerty** [comment]
Seems kind of obtuse to be honest but whatever

File: code/modules/hydroponics/grown/towercap.dm [hunk 7]
```diff
@@ -140,6 +140,8 @@ obj/item/seeds/bamboo
 /obj/structure/punji_sticks/Initialize(mapload)
 	. = ..()
 	AddComponent(/datum/component/caltrop, 20, 30, 100, CALTROP_BYPASS_SHOES)
+	var/turf/T = get_turf(src)
+	usr.investigate_log("has placed punji sticks trap at [AREACOORD(T)].", INVESTIGATE_BOTANY)
```
Comment (PowerfulBacon) [inline code/modules/hydroponics/grown/towercap.dm]: This should really be in the code that places the stick rather an in the sticks init. What about sticks created through game code / admin tools; this should work if they were to be needed.

File: code/modules/hydroponics/plant_genes.dm [hunk 8]
```diff
@@ -253,16 +253,31 @@
...
 /datum/plant_gene/trait/cell_charge/on_squash(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
 	if(iscarbon(target))
 		var/mob/living/carbon/C = target
 		var/power = G.seed.potency*rate
 		if(prob(power))
 			C.electrocute_act(round(power), G, 1, 1)
+			var/offender = G.fingerprintslast
+			if(C.ckey != offender)
+				var/mob/living/carbon/badperson
+				for(var/mob/M in GLOB.mob_living_list)
```
Comment (PowerfulBacon) [inline code/modules/hydroponics/plant_genes.dm]: Get the client by ckey from the associative list and then get their mob from that. (I dont know the lost name but I know it exists)

File: code/modules/hydroponics/plant_genes.dm [hunk 9]
```diff
@@ -330,14 +345,30 @@
...
 		var/teleport_radius = max(round(G.seed.potency / 10), 1)
 		var/turf/T = get_turf(target)
+		var/offender = G.fingerprintslast
+		var/mob/living/carbon/C = target
 		new /obj/effect/decal/cleanable/molten_object(T) //Leave a pile of goo behind for dramatic effect...
 		do_teleport(target, T, teleport_radius, channel = TELEPORT_CHANNEL_BLUESPACE)
+		if(C.ckey == offender)		//what's the point of logging someone attacking himself
+			return
+		var/mob/living/carbon/badperson
+		for(var/mob/M in GLOB.mob_living_list)
```
Comment (PowerfulBacon) [inline code/modules/hydroponics/plant_genes.dm]: Again, see above

File: code/modules/hydroponics/plant_genes.dm [hunk 10]
```diff
@@ -408,7 +439,19 @@
 	name = "Hypodermic Prickles"
 
 /datum/plant_gene/trait/stinging/on_slip(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
-	on_throw_impact(G, target)
+	if(isliving(target) && G.reagents && G.reagents.total_volume)
```
Comment (PowerfulBacon) [inline code/modules/hydroponics/plant_genes.dm]: Could you explain why you aren't just using throw_impact to do the logging?

File: code/modules/hydroponics/plant_genes.dm [hunk 11]
```diff
@@ -419,6 +462,17 @@
 			G.reagents.reaction(L, INJECT, fraction)
 			G.reagents.trans_to(L, injecting_amount)
 			to_chat(target, "<span class='danger'>You are pricked by [G]!</span>")
+			var/offender = G.fingerprintslast
+			var/mob/living/carbon/badperson
+			if(L.ckey != offender)			//what's the point of logging someone attacking himself
+				var/turf/T = get_turf(L)
+				for(var/mob/M in GLOB.mob_living_list)
```
Comment (PowerfulBacon) [inline code/modules/hydroponics/plant_genes.dm]: Again, get client by ckey through the list not a for loop

File: code/modules/hydroponics/plant_genes.dm (hunk 10 above)
Comment (Archanial) [reply code/modules/hydroponics/plant_genes.dm]: Because throw is a throw and it gets logged as combat.

**Comment by Archanial** [comment]
Good to know thrownby exists

**Review by PowerfulBacon — COMMENTED** [review]
Going to be honest, I dont like the repeated code in throw impact and squash, there should be a proc that does the copied code on both and then the log after that

**Comment by Archanial** [comment]
Done

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**